LandSandBoat/server
# Implement Ul'xzomit mother and baby distinction

**_I affirm:_**

## What does this pull request do?

 * Two thirds of the current Ul'xzomit mobs in Al'Taieu are now babies following the remaining one third Ul'xzomits.
 * The baby Ul'xzomit are smaller and the mother Ul'xzomit are larger.
 * Increase roam distance of mother Ul'xzomit.
 * Mother and baby Ul'xzomit have appropriate (different) drop tables, as in: https://ffxiclopedia.fandom.com/wiki/Ul%27xzomit
    * Make elemental clusters slotted drops (10% for mothers, 1% for babies, with equal chance for each cluster).

### Under the hood

 * Create a new lua function in `scripts/globals/follow.lua` that allows you to call `xi.follow.follow(follower, target[, options])` to have the follower mob persistently follow the target entity.
 * Rename `ROAMFLAG_EVENT` to `ROAMFLAG_SCRIPTED` for a more intuitive name.
 * Add a new `MOBMOD_ROAM_RESET_FACING`. This implemented previous behavior that was hard-coded into `ROAMFLAG_EVENT`. Specifically, it makes mobs reface their original spawn facing when they roamed back home (like after the players died).
 * Add a new `MOBMOD_DONT_ROAM_HOME`. If used, it disables the leash that causes mobs to eventually roam back home.
 * Fix some bugs involving `LookAt()` that cropped up while working on this.
 * Add several utility functions in lua and cpp, and modify the `distance` functions to add an optional `ignoreVertical` flag.
 * Rework the lua `setMobFlags()` function to work from a mob as well, like `mob:setMobFlags(flags)`.
 * Adds a few more `triggerListener()` calls for pathing and roaming.

### Not implemented/Need more research

 * Ul'xzomit roam patterns. At the moment they roam normally but with a larger range. Need more retail research.
 * Babies respawning. At the moment they only spawn when the mother spawns. Need more retail research.
 * Baby roam behavior if mother is dead. At the moment they don't roam at all. Need more retail research.
 * Difference in HP/stats of mother vs. baby. Need more retail research.

Fixes #2604 
Fixes #2573 

~~We might consider making the elemental clusters be slotted drops. At the moment they're implemented as 10% and 1% each for the Ul'xzomit mothers and babies, but I think it's supposed to be 10% and 1% that a cluster drops at all. As is, the mothers have a 57.0% chance for at least one cluster dropping and the babies have a 7.7% chance for at least one cluster.~~

Also, is there interest in a gm command for making a mob follow an entity? I'm not sure if that belongs in LSB core.

## Steps to test these changes

1. Go to Al'Taieu.
2. Watch the Ul'xzomit roam about. The babies follow the mothers.
3. Attack a baby Ul'xzomit. Nothing should link.
4. Attack a mother Ul'xzomit. The two babies should link, but no others.
5. Kill a few mother and baby Ul'xzomit. Notice the different drop rates.
